### PR TITLE
fix(types): add undefined initial value to Atom definition

### LINKED ISSUE #2666: [TypeScript] Add `undefined` to Atom definition
### Actual
With:
```ts
const myAtom = atom<string>()
```
In VSCode I get:
```console
Expected 1 arguments, but got 0.ts(2554)
atom.d.ts(38, 37): An argument for 'read' was not provided
```

---

Instead, with:
```ts
const myAtom = atom<string | undefined>()
```

With ESLint I get:
```console
error  Do not use useless `undefined`  unicorn/no-useless-undefined
```

### Expected
No Error in VSCode with:
```ts
const myAtom = atom<string>()
```

### Related example
```ts
const [myState, setMyState] = useState<string>()
```

### PR DESCRIPTION
**Pull request by rtritto**
## Related Bug Reports or Discussions

Discussion #2667

Fixes #2666

## Check List

- [x] `pnpm run prettier` for formatting code and docs

### CODE REVIEW

File: src/vanilla/atom.ts [hunk 1]
```diff
@@ -75,6 +75,11 @@ export function atom<Value, Args extends unknown[], Result>(
 // read-only derived atom
 export function atom<Value>(read: Read<Value>): Atom<Value>
 
+// Value with default as initial value
+export function atom<Value = undefined>(
+  read: Read<Value | undefined>,
+): Atom<Value | undefined>
```
Comment (dai-shi) [inline src/vanilla/atom.ts:81]: Hmm, I was expecting something like this:
```suggestion
export function atom<Value>(): PrimitiveAtom<Value | undefined>
```

Comment (rtritto) [reply src/vanilla/atom.ts:81]: This works but, in `/src/vanilla/atom.ts`, I get
```console
This overload signature is not compatible with its implementation signature.ts(2394)
atom.ts(92, 17): The implementation signature is declared here.
```
The signature (parameter) needs to be added.

Comment (dai-shi) [reply src/vanilla/atom.ts:81]: Yeah, let's do that. Just make it optional?

Comment (rtritto) [reply src/vanilla/atom.ts]: Done

**Comment by rtritto** [comment]
Does this
```diff
// write-only derived atom
export function atom<Value, Args extends unknown[], Result>(
-  initialValue: Value,
+  initialValue?: Value,
  write: Write<Args, Result>,
): WritableAtom<Value | undefined, Args, Result> & WithInitialValue<Value | undefined>
```
also be need a change?

**Comment by dai-shi** [comment]
No, I think it should be explicit, and also TS doesn't allow it.

File: src/vanilla/atom.ts [hunk 2]
```diff
@@ -83,8 +83,8 @@ export function atom<Value, Args extends unknown[], Result>(
 
 // primitive atom
 export function atom<Value>(
-  initialValue: Value,
-): PrimitiveAtom<Value> & WithInitialValue<Value>
+  initialValue?: Value,
+): PrimitiveAtom<Value | undefined> & WithInitialValue<Value | undefined>
```
Comment (dai-shi) [inline src/vanilla/atom.ts:87]: No, it's looks bad. We need overloads.

```ts
// primitive atom without default value
export function atom<Value>(): PrimitiveAtom<Value | undefined> & WithInitialValue<Value | undefined>

// primitive atom
export function atom<Value>(
  initialValue: Value,
): PrimitiveAtom<Value> & WithInitialValue<Value>
```

**Review by rtritto — COMMENTED** [review]
_

File: src/vanilla/atom.ts [hunk 3]
```diff
@@ -26,7 +26,7 @@ type Write<Args extends unknown[], Result> = (
 // This is an internal type and not part of public API.
 // Do not depend on it as it can change without notice.
 type WithInitialValue<Value> = {
-  init: Value
+  init: Value | undefined
```
Comment (dai-shi) [inline src/vanilla/atom.ts:29]: Please don't do this.

File: src/vanilla/atom.ts [hunk 4]
```diff
@@ -86,14 +86,17 @@ export function atom<Value>(
   initialValue: Value,
 ): PrimitiveAtom<Value> & WithInitialValue<Value>
 
+// primitive atom without default value
+export function atom<Value>(): PrimitiveAtom<Value> & WithInitialValue<Value>
+
```
Comment (dai-shi) [inline src/vanilla/atom.ts:91]: Just for preference, I would like to have it before `// primitive atom`.

**Review by dai-shi — CHANGES_REQUESTED** [review]
Sorry if it's confusing, but do not change existing public types. Only add a new overload type and possibly fix our type in the implementation (which is not public).

**Comment by rtritto** [comment]
Done

**Review by dai-shi — COMMENTED** [review]
The code looks good to me.
Can you add some tests?

**Comment by rtritto** [comment]
Done

File: tests/vanilla/basic.test.tsx [hunk 5]
```diff
@@ -2,6 +2,8 @@ import { expect, it } from 'vitest'
 import { atom } from 'jotai/vanilla'
 
 it('creates atoms', () => {
+  // primitive atom without initial value
+  const undefinedAtom = atom()
```
Comment (dai-shi) [inline tests/vanilla/basic.test.tsx:6]: I'm sorry, but I didn't mean this test.
Please revert it and add "type" tests in `tests/vanilla/types.test.tsx`.

**Comment by rtritto** [comment]
@dai-shi I need some help

**Comment by dai-shi** [comment]
Sure, but for which one? Let's first focus on the first one.

**Comment by rtritto** [comment]
It's related to `vanilla/types.test`

**Comment by dai-shi** [comment]
Please add this and if you have more ideas, please add them.

```diff
diff --git a/tests/vanilla/types.test.tsx b/tests/vanilla/types.test.tsx
index e0744b8..3c89725 100644
--- a/tests/vanilla/types.test.tsx
+++ b/tests/vanilla/types.test.tsx
@@ -1,4 +1,5 @@
 import { expectType } from 'ts-expect'
+import type { TypeOf } from 'ts-expect'
 import { it } from 'vitest'
 import { atom } from 'jotai/vanilla'
 import type {
@@ -15,6 +16,15 @@ it('atom() should return the correct types', () => {
     // primitive atom
     const primitiveAtom = atom(0)
     expectType<PrimitiveAtom<number>>(primitiveAtom)
+    expectType<TypeOf<PrimitiveAtom<number>, typeof primitiveAtom>>(true)
+    expectType<TypeOf<PrimitiveAtom<number | undefined>, typeof primitiveAtom>>(
+      false,
+    )
+
+    // primitive atom without initial value
+    const primitiveWithoutInitialAtom = atom<number | undefined>()
+    expectType<PrimitiveAtom<number | undefined>>(primitiveWithoutInitialAtom)
+    expectType<PrimitiveAtom<undefined>>(atom())
 
     // read-only derived atom
     const readonlyDerivedAtom = atom((get) => get(primitiveAtom) * 2)
```

**Review by dai-shi — APPROVED** [review]
LGTM
Thanks for your contribution.

**Comment by dai-shi** [comment]
**Size Change:** 0 B 

**Total Size:** 87 kB



<details><summary>ℹ️ <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| `./dist/babel/plugin-debug-label.js` | 932 B |
| `./dist/babel/plugin-react-refresh.js` | 1.14 kB |
| `./dist/babel/preset.js` | 1.41 kB |
| `./dist/esm/babel/plugin-debug-label.mjs` | 1 kB |
| `./dist/esm/babel/plugin-react-refresh.mjs` | 1.19 kB |
| `./dist/esm/babel/preset.mjs` | 1.49 kB |
| `./dist/esm/index.mjs` | 62 B |
| `./dist/esm/react.mjs` | 1.02 kB |
| `./dist/esm/react/utils.mjs` | 746 B |
| `./dist/esm/utils.mjs` | 67 B |
| `./dist/esm/vanilla.mjs` | 3.83 kB |
| `./dist/esm/vanilla/utils.mjs` | 4.95 kB |
| `./dist/index.js` | 242 B |
| `./dist/react.js` | 1.06 kB |
| `./dist/react/utils.js` | 1.39 kB |
| `./dist/system/babel/plugin-debug-label.development.js` | 1.1 kB |
| `./dist/system/babel/plugin-debug-label.production.js` | 775 B |
| `./dist/system/babel/plugin-react-refresh.development.js` | 1.29 kB |
| `./dist/system/babel/plugin-react-refresh.production.js` | 928 B |
| `./dist/system/babel/preset.development.js` | 1.59 kB |
| `./dist/system/babel/preset.production.js` | 1.14 kB |
| `./dist/system/index.development.js` | 252 B |
| `./dist/system/index.production.js` | 183 B |
| `./dist/system/react.development.js` | 1.17 kB |
| `./dist/system/react.production.js` | 715 B |
| `./dist/system/react/utils.development.js` | 860 B |
| `./dist/system/react/utils.production.js` | 462 B |
| `./dist/system/utils.development.js` | 257 B |
| `./dist/system/utils.production.js` | 187 B |
| `./dist/system/vanilla.development.js` | 3.92 kB |
| `./dist/system/vanilla.production.js` | 2.06 kB |
| `./dist/system/vanilla/utils.development.js` | 5.16 kB |
| `./dist/system/vanilla/utils.production.js` | 3.09 kB |
| `./dist/umd/babel/plugin-debug-label.development.js` | 1.08 kB |
| `./dist/umd/babel/plugin-debug-label.production.js` | 852 B |
| `./dist/umd/babel/plugin-react-refresh.development.js` | 1.27 kB |
| `./dist/umd/babel/plugin-react-refresh.production.js` | 1 kB |
| `./dist/umd/babel/preset.development.js` | 1.54 kB |
| `./dist/umd/babel/preset.production.js` | 1.22 kB |
| `./dist/umd/index.development.js` | 383 B |
| `./dist/umd/index.production.js` | 328 B |
| `./dist/umd/react.development.js` | 1.18 kB |
| `./dist/umd/react.production.js` | 786 B |
| `./dist/umd/react/utils.development.js` | 1.53 kB |
| `./dist/umd/react/utils.production.js` | 1.01 kB |
| `./dist/umd/utils.development.js` | 399 B |
| `./dist/umd/utils.production.js` | 342 B |
| `./dist/umd/vanilla.development.js` | 4.91 kB |
| `./dist/umd/vanilla.production.js` | 2.69 kB |
| `./dist/umd/vanilla/utils.development.js` | 6.07 kB |
| `./dist/umd/vanilla/utils.production.js` | 3.72 kB |
| `./dist/utils.js` | 247 B |
| `./dist/vanilla.js` | 4.81 kB |
| `./dist/vanilla/utils.js` | 5.94 kB |

</details>



<a href="https://github.com/preactjs/compressed-size-action"><sub>compressed-size-action</sub></a>

**Pull request merged**